File: gateway/src/main/java/com/beat/gateway/security/internal/servlet/SecurityMdcLoggingFilter.java
package com.beat.gateway.security.internal.servlet;

import com.beat.observability.logging.filter.BaseMdcLoggingFilter;
import com.beat.observability.tracing.TraceContextResolver;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.security.authentication.AnonymousAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;

public class SecurityMdcLoggingFilter extends BaseMdcLoggingFilter {

	private final int managementPort;

	public SecurityMdcLoggingFilter(TraceContextResolver traceContextResolver) {
		this(traceContextResolver, -1);
	}

	public SecurityMdcLoggingFilter(TraceContextResolver traceContextResolver, int managementPort) {
		super(traceContextResolver);
		this.managementPort = managementPort;
	}

	@Override
	protected boolean shouldNotFilter(HttpServletRequest request) {
		return managementPort > 0 && request.getLocalPort() == managementPort;
	}

	@Override
	protected String resolveUserId() {
		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
		if (authentication == null || authentication instanceof AnonymousAuthenticationToken) {
			return null;
		}

		Object principal = authentication.getPrincipal();
		if (principal instanceof Long memberId) {
			return memberId.toString();
		}

		String name = authentication.getName();
		return name == null || name.isBlank() ? null : name;
	}
}

File: gateway/src/main/kotlin/com/beat/support/security/authentication/internal/SecurityMdcLoggingFilter.kt
package com.beat.support.security.authentication.internal

import com.beat.observability.logging.filter.BaseMdcLoggingFilter
import com.beat.observability.tracing.TraceContextResolver
import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.authentication.AnonymousAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder

/**
 * SecurityContext의 principal을 MDC `userId`로 노출한다.
 * management port 요청은 access log/MDC 대상에서 제외한다.
 */
internal class SecurityMdcLoggingFilter constructor(
    traceContextResolver: TraceContextResolver,
    private val managementPort: Int = NO_MANAGEMENT_PORT,
    actuatorBasePath: String = "/actuator",
) : BaseMdcLoggingFilter(traceContextResolver, actuatorBasePath) {

    override fun shouldNotFilter(request: HttpServletRequest): Boolean =
        managementPort > 0 && request.localPort == managementPort

    protected override fun resolveUserId(): String? {
        val authentication = SecurityContextHolder.getContext().authentication
            ?.takeIf { it !is AnonymousAuthenticationToken }
            ?: return null

        return (authentication.principal as? Long)?.toString()
            ?: authentication.name?.takeIf(String::isNotBlank)
    }

    companion object {
        private const val NO_MANAGEMENT_PORT = -1
    }
}
